### PR TITLE
Assert that layout construction from extents is valid where used

### PR DESCRIPTION
**Pull request by crtrott**
Also introduces a missing layout overload for required_allocation_size

### CODE REVIEW

**Comment by crtrott** [comment]
Fixes #5205

File: core/src/Kokkos_View.hpp [hunk 1]
```diff
@@ -1536,7 +1536,11 @@ class View : public ViewTraits<DataType, Properties...> {
       : View(arg_prop,
              typename traits::array_layout(arg_N0, arg_N1, arg_N2, arg_N3,
                                            arg_N4, arg_N5, arg_N6, arg_N7),
-             check_input_args::yes) {}
+             check_input_args::yes) {
+    static_assert(traits::array_layout::is_extent_constructible,
+                  "Layout is not extent constructible. Use overload taking a "
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:1541]: Please suggest a message that will be easier to understand for users

Comment (crtrott) [reply core/src/Kokkos_View.hpp:1541]: I don't know, seems pretty obvious to me?

Comment (dalg24) [reply core/src/Kokkos_View.hpp:1541]: "is not constructible from extent arguments.  Use ..."

Comment (masterleinad) [reply core/src/Kokkos_View.hpp:1541]: I would just use the second sentence "Use the overload taking a layout object instead!".

**Review by nliber — APPROVED** [review]

**Pull request merged**